### DISCUSSION TITLE
Try install homebrew on arm64 ubuntu

### DISCUSSION THREAD

**Discussion by zouchengli** [Homebrew/discussions#4796] [Linux]
### Additional Info

- [X] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
lima@lima-ubuntu-lts:~$ ruby -v
ruby 2.6.10p210 (2022-04-12 revision 67958) [aarch64-linux]
```

### Output of `brew doctor`

```shell
lima@lima-ubuntu-lts:~$ echo $PATH
/home/lima.linux/.rbenv/shims:/home/lima.linux/.rbenv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/sbin:/sbin:/usr/sbin:/sbin:/usr/sbin:/sbin:/home/lima.linux/.rbenv/versions/2.6.10/bin/
```

### Description of issue

lima@lima-ubuntu-lts:~$ ./install.sh
==> Checking for sudo access (which may request your password)...
==> This script will install:
/home/linuxbrew/.linuxbrew/bin/brew
/home/linuxbrew/.linuxbrew/share/doc/homebrew
/home/linuxbrew/.linuxbrew/share/man/man1/brew.1
/home/linuxbrew/.linuxbrew/share/zsh/site-functions/_brew
/home/linuxbrew/.linuxbrew/etc/bash_completion.d/brew
/home/linuxbrew/.linuxbrew/Homebrew

Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /bin/chown -R lima:lima /home/linuxbrew/.linuxbrew/Homebrew
==> Downloading and installing Homebrew...
HEAD is now at https://github.com/Homebrew/brew/commit/aa2a77b1f1ac1539e69e186bad1d569100448987 Merge pull request https://github.com/Homebrew/brew/pull/16010 from apainintheneck/improve-versioned-formula-names-performance
Error: No Homebrew ruby 2.6.10_1 available for aarch64 processors!
Error: Failed to install Homebrew Portable Ruby and cannot find another Ruby 2.6.10!
If there's no Homebrew Portable Ruby available for your processor:

install Ruby 2.6.10 with your system package manager (or rbenv/ruby-build)
make it first in your PATH
try again
Failed during: /home/linuxbrew/.linuxbrew/bin/brew update --force --quiet

**Comment by SMillerDev** [comment]
https://docs.brew.sh/Homebrew-on-Linux#arm-unsupported

**Reply by zouchengli** [reply]
I changed the script to automatically identify the chip architecture. I want to try to install the corresponding ruby version to drive homebrew. Isn't this possible?

**Reply by SMillerDev** [reply]
Since ARM linux is unsupported there is no Homebrew pre-built ruby for it. The options for installing the ruby versions are explained in the error:
```
install Ruby 2.6.10 with your system package manager (or rbenv/ruby-build)
make it first in your PATH
```

**Reply by naikrovek** [reply]
Is there a root reason that linux/arm64 is unsupported?  It seems to me (an outsider, without context) that linux/arm64 isn't supported solely because it isn't supported.

Is there anything that one can do to help make homebrew on arm64 Linux a supported architecture?

**Reply by SMillerDev** [reply]
Because there is no CI and nobody willing to put in the work

**Reply by naikrovek** [reply]
what work is involved?  Again, I'm new to homebrew, so sorry for asking a silly question.  Things need to be compiled into bottles I assume, and maintained, right?  what else?

**Reply by SMillerDev** [reply]
All the code in brew needs to be adjusted to account for an ARM Linux platform, all ~6000 pieces of software that homebrew ships need to be adjusted to work on Arm for Linux and all CI needs to be adjusted to actually test this combination. After that CI can start compiling bottles and the community can then maintain those bottles, yes.

**Reply by naikrovek** [reply]
seems like the number of changes would be very small.  I've written and compiled plenty of software for Windows and MacOS and Linux and none of the platform specific stuff was insurmountable.

Linux on amd64 versus Linux on arm64 seems (seems) like a very small change compared to what I've done in the past.  the _number_ of packages here, though.

95% will require no work, 4.9% will require changes to how they are built, and 0.1% will be a major pain in the tochus.

I'll look more closely, thank you.  I think you were intending to convince me that it is impossible, and it definitely isn't.

**Reply by SMillerDev** [reply]
It's not impossible, but it is a lot of work that nobody is currently doing. So it's not as simple as just bottling things. 

For starters, Linux on ARM is not a platform that GitHub Actions (where Homebrews CI runs) supports. So you can write the code, but without a way to validate it it'll be a hard sell to officially support.

**Reply by naikrovek** [reply]
ok _that_ is a good reason that this platform & architecture combination are not supported.  I can stand up some single-use virtual machines on my Mac for my own stuff, but obviously that will not scale.

**Reply by naikrovek** [reply]
I wonder if lack of support here is holding up the implementation of arm64 Linux GitHub Actions runners.  They use Homebrew on the Intel Linux runner images...  That would be kinda funny.

can cross-compilation be done on amd64 Linux runners, to get started?  I haven't done cross-compilation in ... a while, so I don't know how easy or difficult this is, anymore.  it used to be something of a pain.

**Reply by SMillerDev** [reply]
Homebrew never does cross compilation anymore since it only ever brought us problems. Builds are optimised for the hardware they were made on.

I'm sure GitHub would ship runners without brew if that was the limiting factor.

**Reply by naikrovek** [reply]
you're right.  there was an announcement at GitHub Universe.  

https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/

**Comment by ZhongRuoyu** [comment]
What you encountered was likely an issue running `brew update-report` at the end of `brew update`. When portable-ruby is unavailable, `brew` tries to find a working `ruby` from `HOMEBREW_PATH` [^1], the unfiltered `PATH` from the environment [^2]. However, when `brew` is recursively invoked as a subprocess of itself, the already-filtered `PATH` from the parent `brew` process becomes the `HOMEBREW_PATH` in the subprocess. Having lost the original, unfiltered `PATH` information, `brew update-report`, run as a subprocess of `brew` [^3], failed to find a working `ruby`, although it should be on `PATH` in the outside environment.

While running `brew` on aach64 Linux is still unsupported at the moment, you are welcome to make pull requests to improve the current situation. Or, if you just want to get it work for now, maybe try doing this somewhere in your modified installer script: https://github.com/ZhongRuoyu/homebrew-aarch64-linux/commit/dcd91dc8cb6f3f5591ee3229284ec5c6fca042eb. By linking the rbenv-installed Ruby to where portable-ruby is expected, this effectively tricks `brew` by pretending that a portable-ruby is available. (This is not supported, so it can break anytime in the future.) If you do that, you may also want to remove rbenv's shims (`~/.rbenv/shims`) from your `PATH`, since `brew cleanup` may consider the "portable-ruby" unnecessary and proceed to remove it. (Try `brew cleanup --dry-run` to see it yourself.)

[^1]: https://github.com/Homebrew/brew/blob/1f80e82dd72fce6642e4ea360f7a1feef5d78470/Library/Homebrew/utils/ruby.sh#L62
[^2]: https://github.com/Homebrew/brew/blob/1f80e82dd72fce6642e4ea360f7a1feef5d78470/bin/brew#L152-L168
[^3]: https://github.com/Homebrew/brew/blob/1f80e82dd72fce6642e4ea360f7a1feef5d78470/Library/Homebrew/cmd/update.sh#L896

**Comment by ssbarnea** [comment]
Any chance to have an update on this? We had ubuntu arm64 github runners for more than an year, so is not that not we cannot really test it.

**Reply by SMillerDev** [reply]
It's only been a week for public repos.

EDIT: https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/
